### DISCUSSION TITLE
Copilot documentation for plugin developers

### DISCUSSION THREAD

**Discussion by zbirenbaum** [community/community#13619] [Copilot Conversations]
I’m a neovim user that recently got access and am trying to write a copilot plugin in lua, as the currently provided copilot.vim is entirely vimscript, making it impossible to integrate with modern neovim plugins and very difficult to customize. The plugin as is interferes greatly with my existing workflow and configurations, but it often gives really good suggestions so I would hate to have to disable it. 

Unfortunately, while I managed to get the server running remotely and logs show that the requests I send to it (which I built by directly porting the vimscript which makes the requests and printing it to make sure everything  is identical) I never seem to get a server response unless there’s an error. I know it is working, because if I send a non existent method like “init” instead of “initialize” I get an error from my error callback, and if I send incorrect parameters for a method, I get a different error from the same callback accurately stating the problem.  When I send correct methods and parameters though, it’s radio silent.

The vimscript is not well documented, likely because it was never intended to be ported or built upon, and the copilot agent seems to be a modified lsp server, but there doesn’t seem to be any way to get the arguments/requirements/handlers from docs or help files, and @tpope has disabled all issues/discussions so I can’t really ask there (which I honestly don’t blame him for, considering the number of posts here that are just complaints about waitlist). 

Since the requests are properly being read from stdin, I can’t tell if it’s an issue with my lua authentication code failing silently, or some way of interfacing with the process stdout.

Could one of the developers provide some help here, ideally by shedding some light on the relevant parts of the agent for interfacing with it?

**Comment by tpope** [comment]
Big picture, stay tuned, changes are coming. I would discourage you from wasting your time trying to reverse engineer in the meantime.

> I’m a neovim user that recently got access and am trying to write a copilot plugin in lua, as the currently provided copilot.vim is entirely vimscript, making it impossible to integrate with modern neovim plugins and very difficult to customize. The plugin as is interferes greatly with my existing workflow and configurations, but it often gives really good suggestions so I would hate to have to disable it.

Do feel free to elaborate on how it interferes and how you envision integrating.

**Reply by zbirenbaum** [reply]
> Big picture, stay tuned, changes are coming. I would discourage you from wasting your time trying to reverse engineer in the meantime.
> 
> > I’m a neovim user that recently got access and am trying to write a copilot plugin in lua, as the currently provided copilot.vim is entirely vimscript, making it impossible to integrate with modern neovim plugins and very difficult to customize. The plugin as is interferes greatly with my existing workflow and configurations, but it often gives really good suggestions so I would hate to have to disable it.
> 
> Do feel free to elaborate on how it interferes and how you envision integrating.

Thank you very much for the response. My personal end goal once getting it into lua was to make copilot into a new cmp source, similar to tabnine, and then open a floating preview with normal language highlighting when I hover the snippet in my completion menu, similar to a documentation preview. 

I believe it would reduce cpu usage a bit since cmp debounces calling it’s sources rather than running on every insert change on top of being more cleanly integrated (right now my existing ghost text interferes and I find the large blocks moving my code around disturbing)

Its possible this could be done from vimscript in a way I’m unaware of, but I’m definitely not good enough at vimscript to implement it.

With that said, it was also one of my goals to provide a generic api so that plugin authors or just other users could access the copilot outputs to integrate in other plugins like statuslines as they see fit.

**Reply by zbirenbaum** [reply]
As a followup for anyone interested, I put together two plugins. One for starting the copilot server using the lsp module in neovim, and another for providing cmp completions, with full previews in the documentation area as described above. Currently I piggyback on copilot.vim's authentication for simplicity's sake (and to make sure the plugin follows the ToS/license) but after releasing it early this week, most of the alpha stage bugs appear to have been ironed out. Anyone interested in a lua api for interacting with copilot or an out of the box solution for using it as a cmp source can find the repositories here:

- https://github.com/zbirenbaum/copilot.lua
- https://github.com/zbirenbaum/copilot-cmp

**Reply by Leefdaniel091** [reply]
https://github.com/github/feedback/discussions/13619#discussioncomment-2534234
